python/shotgun_model/shotgun_query_model.py

```diff
@@ -620,262 +557,271 @@ def _set_tooltip(self, item, data):
+                      can also for example be ``sg_sequence.Sequence.image``.
+        :param url: thumbnail url
+        :param entity_type: Shotgun entity type
+        :param entity_id: Shotgun entity id
+        """
+        if url is None:
+            # nothing to download. bad input. gracefully ignore this request.
+            return
+
+        if not self._sg_data_retriever:
+            raise sgtk.ShotgunModelError("Data retriever is not available!")
+
+        uid = self._sg_data_retriever.request_thumbnail(
+            url,
+            entity_type,
+            entity_id,
+            field,
+            self.__bg_load_thumbs
         )
 
-        try:
-            time_before = time.time()
-            num_items = self.__load_from_disk()
-            time_diff = (time.time() - time_before)
-            self._log_debug(
-                "Loading finished! Loaded %s items in %4fs" %
-                (num_items, time_diff)
-            )
-            self.cache_loaded.emit()
-            return True
-        except Exception, e:
-            self._log_debug(
-                "Couldn't load cache data from disk.\n"
-                " Will proceed with full SG load.\n"
-                "Error reported: %s" % (e,)
-            )
-            return False
+        # keep tabs of this and call out later - note that we use a weakref to allow
+        # the model item to be gc'd if it's removed from the model before the thumb
+        # request completes.
+        self.__thumb_map[uid] = {
+            "item_ref": weakref.ref(item),
+            "field": field
+        }
 
-    def _log_debug(self, msg):
+    def _ensure_item_loaded(self, uid):
         """
-        Convenience wrapper around debug logging
+        Ensures that the given unique id is loaded by the model.
 
-        :param msg: debug message
+        :param str uid: Unique id for a :class:`ShotgunDataHandler` item.
+        :returns: :returns: :class:`~PySide.QtGui.QStandardItem` or ``None`` if not found
         """
-        self._bundle.log_debug("[%s] %s" % (self.__class__.__name__, msg))
+        data_item = self._data_handler.get_data_item_from_uid(uid)
 
-    def _log_warning(self, msg):
-        """
-        Convenience wrapper around warning logging
+        if data_item is None:
+            # no match in data store
+            self._log_debug("...uid '%s' is not part of the data set" % uid)
+            return None
 
-        :param msg: debug message
-        """
-        self._bundle.log_warning("[%s] %s" % (self.__class__.__name__, msg))
+        # this node is loaded in the query cached by the data handler
+        # but may not exist in the model yet - because of deferred loading.
 
-    def _sg_clean_data(self, sg_data):
-        """
-        Recursively clean the supplied SG data for use by clients.
+        # first see if we have it in the model already
+        item = self._get_item_by_unique_id(uid)
+
+        if not item:
```

Comment:
i don't like returning early, it can get confusing too. indent isn't too bad, less than one page long, so fairly readable. i'll leave it as is.